platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,436 @@
+and pass that roster to the consensus system. **This must be deterministic**. Each node in the network must assign the
+same roster to the same rounds. However, Consensus and Execution run in completely different threads, and possibly,
+in different processes.
+
+With this architecture, we introduce a new concept into `EventCore`. `EventCore` has a `payload`, which is a `byte[]`
+that the Execution layer can use to carry whatever payload is meaningful to it. The consensus layer needs an analogous
+section of the `EventCore` into which messages particular to Consensus can be placed. This section will be called
+`repeated control_messages`. A `ControlMessage` is a protobuf message used to control, in some way, the consensus
+system.
+
+TODO: I'm seriously thinking of having "system_transactions" and "user_transactions" instead of "control_messages" and
+"payload".
+
+`ControlMessage` has a one-of, among which is `RosterChangeMessage`. This message will have the new roster, and be
+signed by the network as a whole. Each honest node will submit a `RosterChangeMessage`, and the Hashgraph module will
+recognize these messages. When it encounters the **first** such message for a given roster in a round that has come to
+consensus, the very next round will use that new activated roster.
+
+### Event Creator Module
+
+Every node in the network participating in consensus is permitted to create events to gossip to peers. These events
+are used both for transmitting user transactions, and as the basis of the hashgraph algorithm for "gossiping about
+gossip". Therefore, the Event Creator has two main responsibilities:
+
+1. Create events with "other parent(s)" so as to help the hashgraph progress consensus
+2. Fill events with transactions to be sent to the network through Gossip
+
+#### Creating Events
+
+The Event Creator is configured with a `maximum_event_creation_frequency`, measured in events/sec. This is a network
+wide setting. If any node creates events more rapidly than this setting, then the node will be reported to the Sheriff.
+An event is not necessarily created at this frequency, but will be created at no more than this frequency.
+
+When it is time to potentially create an event, the Event Creator will determine whether it *should* create the event.
+It may consider whether there are any transactions to send, or whether creating an event will help advance the
+hashgraph. It may decide that creating the event would be bad for the network, and veto such creation. Or it may decide
+that creating the event should be permitted.
+
+If the event is to be created, the Event Creator will decide which nodes to select as "other parents". Today, we have
+exactly one "other parent" per event, but multiple "other parents" is shown to effectively reduce latency and network
+traffic. While the implementation of Event Creator may choose to support only a single "other parent", the module is
+designed and intended to support multiple "other parents".
```

Comment:
Had this implemented and passing tests. Branch went stale. 🙃

Reply:
Yes 😭. It's day is coming!